hapi-loader/src/test/java/com/arkhn/hapi/loader/EmbeddedKafkaIntegrationTest.java

```diff
@@ -0,0 +1,150 @@
+    private static PostgreSQLContainer<?> embeddedPostgres;
+
+    @Autowired
+    public KafkaTemplate<String, String> template;
+
+    @Autowired
+    DaoRegistry daoRegistry;
+
+    @Autowired
+    FhirContext myFhirContext;
+
+    @Autowired
+    private KafkaProducer producer;
+
+    @Value("${test.topic}")
+    private String topic;
+
+    private static final String ELASTIC_VERSION = "7.10.1";
+    private static final String ELASTIC_IMAGE = "docker.elastic.co/elasticsearch/elasticsearch:" + ELASTIC_VERSION;
+
+    private static final String POSTGRES_IMAGE = "postgres:13";
+
+    @BeforeAll
+    public static void beforeClass() {
+        embeddedPostgres = new PostgreSQLContainer<>(POSTGRES_IMAGE).withDatabaseName("test-hapi-db").withUsername("sa")
+                .withPassword("sa");
+        embeddedPostgres.start();
+
+        embeddedElastic = new ElasticsearchContainer(ELASTIC_IMAGE)
+                .withStartupTimeout(Duration.of(300, ChronoUnit.SECONDS));
+        embeddedElastic.start();
+    }
+
+    @PreDestroy
+    public void stop() {
+        embeddedPostgres.stop();
+        embeddedElastic.stop();
+    }
+
+    @Test
+    public void loadNewResource() throws Exception {
+        // create a patient
+        Patient p = new Patient();
+        p.setId("loadNewResource");
+        p.setGender(AdministrativeGender.FEMALE);
+
+        // serialize it and send it to the loader
+        IParser parser = myFhirContext.newJsonParser();
+        String serialized = parser.encodeResourceToString(p);
+        producer.send(topic, serialized);
+
+        // wait a bit (could use a refactor)
+        Thread.sleep(3000);
+
+        // assert it has been created in the db
+        IFhirResourceDao<Patient> dao = daoRegistry.getResourceDao("Patient");
+        Patient result = dao.read(p.getIdElement());
+        assertNotNull(result);
+    }
+
+    @Test
+    public void loadExistingResource() throws Exception {
+        IFhirResourceDao<Patient> dao = daoRegistry.getResourceDao("Patient");
+
+        // create initial patient
+        Patient p = new Patient();
+        p.setGender(AdministrativeGender.FEMALE);
+        DaoMethodOutcome outcome = dao.create(p);
+        Patient created = (Patient) outcome.getResource();
+
+        // modify it and send it to the loader
+        Patient updated = new Patient();
+        updated.setId(created.getIdElement());
+        updated.setGender(AdministrativeGender.MALE);
+
+        IParser parser = myFhirContext.newJsonParser();
+        String serialized = parser.encodeResourceToString(updated);
+        producer.send(topic, serialized);
+
+        // wait a bit (could use a refactor)
+        Thread.sleep(3000);
+
+        // assert it has been modified in the db
+        // NB: make sure we remove the "version" part of the ID, otherwise a specific
+        // version of the resource is read
+        Patient result = dao.read(updated.getIdElement().toVersionless());
```

Comment:
Is `toVersionless` needed?

Reply:
yes, otherwise we retrieve a specific version of the resource